File: leetcode/medium/0006-zigzag-conversion/solution.kt
class Solution {
    fun convert(s: String, numRows: Int): String {
        if (numRows == 1 || numRows >= s.length) {
            return s
        }

        val rows = Array(numRows) { StringBuilder() }

        var currentRow = 0
        var direction = -1

        for (ch in s) {
            rows[currentRow].append(ch)

            // Change direction at top and bottom rows
            if (currentRow == 0 || currentRow == numRows - 1) {
                direction *= -1
            }

            currentRow += direction
        }

        val result = StringBuilder()

        for (row in rows) {
            result.append(row)
        }

        return result.toString()
    }
}

File: leetcode/medium/0006-zigzag-conversion/solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s

        rows = [""] * numRows
        current_row = 0
        direction = -1

        for ch in s:
            rows[current_row] += ch

            # Change direction at top and bottom
            if current_row == 0 or current_row == numRows - 1:
                direction *= -1

            current_row += direction

        return "".join(rows)